zeroize/src/lib.rs

```diff
@@ -297,7 +297,21 @@ where
 {
     fn zeroize(&mut self) {
         match self {
-            Some(value) => value.zeroize(),
+            Some(value) => {
+                value.zeroize();
+                // Ensures self is None and that the value was dropped.
+                self.take();
+                // Ensures self is overwritten with the default bit pattern. volatile_write can't be
+                // used because Option<Z> is not copy.
+                //
+                // Safety:
+                //
+                // self is safe to replace with the default, which the take() call above should have
+                // already done semantically. Any value which needed to be dropped will have been
+                // done so by take().
+                unsafe { ptr::write_volatile(self, Option::default()) }
+                atomic_fence();
```

Comment:
I'm unclear on what this is trying to accomplish.

I agree adding the `self.take()` makes sense, but at this point `value` has already been zeroized.

Reply:
I am trying to zeroize the discriminant of the `Option` enum. Logically, `self.take()` will do that, but I copied the approach of "actually, really" zeroing it out as that is the purpose of this crate.

Reply:
In that case the `take` seems redundant

Reply:
Without the take, the drop of the (zeroized) value isn't called. I acknowledge that it's a bit of an edge case considering a zeroized value probably doesn't have anything interesting going on in its drop impl. However, zeroizing `Vec` doesn't free the underlying memory, so there would be a leak if zeroize were called on `Option<Vec<T>>` without the `take`. I am basing this off the documentation of `write_volatile`: 

> `write_volatile` does not drop the contents of `dst`. This is safe, but it could leak allocations or resources, so care should be taken not to overwrite an object that should be dropped.

Additionally, the implementation for the `Zeroize` trait is up to the clients of this crate, who might have their own reasons for expecting a drop call after their type is zeroized. This solution presents the least surprise to the users because the drop is still called, and the in-memory traces of whether the `Option` was `None` or `Some` are zeroized.

Reply:
Aah yes, indeed... it would be a leak.

I think it could probably use some comments to that effect, and the combination of a take and a compiler fence would seem sufficient to ensure the write to the discriminant isn't elided. But belt and suspenders, I suppose...

Reply:
I thought the `write_volatile` was mandatory based on the crate's documentation.

> This crate guarantees the following:
> 1. The zeroing operation can't be "optimized away" by the compiler
> ...
>
> LLVM's volatile semantics ensure 1 is true.

I'm happy to leave it out if you think it's redundant.

Reply:
I think it's okay to do both, with comments that reflect the rationale.

I suppose I'm a bit dubious as to a threat model where a 1-bit discriminant is considered secret information, especially considering all accesses branch upon this value. But I suppose it's worth zeroing out like everything else.

Reply:
I amended the comments and force pushed. Is what I have now adequate?

Reply:
Looks good, thank you!